Register operators of CV models in PyTorch mobile

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27554 [Lite Interpreter] Do not land, test models
* **#27379 Register operators of CV models in PyTorch mobile**

Currently the operators need to be registered manually through c10 registration. 

Tests:
The operators should be covered by tests on operators.
A few ops (add, conv) are covered in test_lite_interpreter.cpp for demonstration. 
CV models may be too large to include in unittests.
Simple local loaders can be built. Follow similar pattern as in test_lite_interpreter to
1. load the torch script model
2. run the model to get reference results
3. save and load the mobile module using torch::jit::module._save_for_mobile() and torch::jit::_load_for_mobile().
4. run the mobile module by run_method() and compare the results to reference results. 

Tested models:
* Lenet
* XrayMobileV3

Differential Revision: [D17810912](https://our.internmc.facebook.com/intern/diff/D17810912)

Commit:
Register operators of CV models in PyTorch mobile

ljk53 commented on torch/csrc/jit/mobile/register_mobile_ops.cpp:
```diff
@@ -24,15 +24,132 @@ static auto registry0 = torch::RegisterOperators().op(
+"_aten::_convolution",
+torch::RegisterOperators::options().kernel(c10::TensorTypeId::CPUTensorId,
+                                           [](at::Tensor input, at::Tensor weight, c10::optional<at::Tensor> bias,
+                                              std::vector<int64_t> stride, std::vector<int64_t> padding,
```
are we always passing these std::vector<> by value instead of by reference? what's the common practice here?
Tensor is fine as they are just pointers.

iseeyuan replied:
Good catch. I'm thinking of c10::ArrayRef, but it has compile error "error: no type named 'element_type' in 'c10::ArrayRef<long long>'". @smessmer do you know why it does not work? I'm fixing it using another solution, the boxed op registration.

ljk53 commented on torch/csrc/jit/mobile/register_mobile_ops.cpp:
```diff
@@ -24,15 +24,132 @@ static auto registry0 = torch::RegisterOperators().op(
-  })
+"_aten::add_.Tensor",
+torch::RegisterOperators::options().kernel(c10::TensorTypeId::CPUTensorId,
+                                           [](at::Tensor a, at::Tensor b, at::Scalar c) ->at::Tensor {
```
nit: space after '->' (I did a search in the codebase and it seems to be common style to have space on both sides)

ljk53 commented on torch/csrc/jit/mobile/register_mobile_ops.cpp:
```diff
@@ -24,15 +24,132 @@ static auto registry0 = torch::RegisterOperators().op(
-  return at::_convolution(input, weight, optional_to_tensor(bias), stride, padding, dilation,
-                     transposed, output_padding, groups, benchmark, deterministic, cudnn_enabled);
-  })
+"_aten::add_.Tensor",
```
nit: indent inside the '.op(' block?

Commit:
Update on "Register operators of CV models in PyTorch mobile"


Currently the operators need to be registered manually through c10 registration. 

Tests:
The operators should be covered by tests on operators.
A few ops (add, conv) are covered in test_lite_interpreter.cpp for demonstration. 
CV models may be too large to include in unittests.
Simple local loaders can be built. Follow similar pattern as in test_lite_interpreter to
1. load the torch script model
2. run the model to get reference results
3. save and load the mobile module using torch::jit::module._save_for_mobile() and torch::jit::_load_for_mobile().
4. run the mobile module by run_method() and compare the results to reference results. 

Tested models:
* Lenet
* XrayMobileV3

ljk53 commented on torch/csrc/jit/mobile/register_mobile_ops.cpp:
```diff
@@ -7,31 +7,219 @@ using torch::jit::peek;
+         (std::move(peek(*stack, 9, 12))).toBool(),
+         (std::move(peek(*stack, 10, 12))).toBool(),
+         (std::move(peek(*stack, 11, 12))).toBool()
+         );
```
chatted offline - I think it's a bit redundant to manually write these lines, but it's fine if this is just temporary as it's going to be replaced by aten->c10 migration work in future.

Commit:
Update on "Register operators of CV models in PyTorch mobile"


Currently the operators need to be registered manually through c10 registration. 

Tests:
The operators should be covered by tests on operators.
A few ops (add, conv) are covered in test_lite_interpreter.cpp for demonstration. 
CV models may be too large to include in unittests.
Simple local loaders can be built. Follow similar pattern as in test_lite_interpreter to
1. load the torch script model
2. run the model to get reference results
3. save and load the mobile module using torch::jit::module._save_for_mobile() and torch::jit::_load_for_mobile().
4. run the mobile module by run_method() and compare the results to reference results. 

Tested models:
* Lenet
* XrayMobileV3

Differential Revision: [D17810912](https://our.internmc.facebook.com/intern/diff/D17810912)

peterjc123 commented:
Windows CI is failing. cc@ezyang
Error log:
```
18:58:17 FAILED: caffe2/CMakeFiles/torch.dir/__/torch/csrc/jit/mobile/register_mobile_ops.cpp.obj 
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\build\win_tmp\bin\sccache.exe  cl   /TP -DAT_PARALLEL_OPENMP=1 -DCAFFE2_BUILD_MAIN_LIB -DCPUINFO_SUPPORTED_PLATFORM=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTH_BLAS_MKL -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_DEPRECATE=1 -D_OPENMP_NOFORCE_MANIFEST -D_THP_CORE -Dtorch_EXPORTS -Iaten\src -I..\aten\src -I. -I..\ -I..\cmake\..\third_party\benchmark\include -Icaffe2\contrib\aten -I..\third_party\onnx -Ithird_party\onnx -I..\third_party\foxi -Ithird_party\foxi -I..\caffe2\..\torch\csrc\api -I..\caffe2\..\torch\csrc\api\include -I..\caffe2\aten\src\TH -Icaffe2\aten\src\TH -I..\caffe2\..\torch\..\aten\src -Icaffe2\aten\src -Icaffe2\..\aten\src -Icaffe2\..\aten\src\ATen -I..\caffe2\..\torch\csrc -I..\caffe2\..\torch\..\third_party\miniz-2.0.8 -I..\aten\src\TH -I..\aten\..\third_party\catch\single_include -I..\aten\src\ATen\.. -Icaffe2\aten\src\ATen -I..\third_party\miniz-2.0.8 -I..\caffe2\core\nomnigraph\include -I..\c10\.. -I..\third_party\cpuinfo\include -I..\third_party\FP16\include -I..\cmake\..\third_party\googletest\googlemock\include -I..\cmake\..\third_party\googletest\googletest\include -I..\third_party\protobuf\src -Iwin_tmp\mkl\include -I..\cmake\..\third_party\eigen -IC:\Jenkins\Miniconda3\include -IC:\Jenkins\Miniconda3\lib\site-packages\numpy\core\include -I..\cmake\..\third_party\pybind11\include -I\opt\rocm\hip\include -I\include -I..\caffe2 /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD /Z7 /EHa /DNOMINMAX /wd4267 /wd4251 /wd4522 /wd4838 /wd4305 /wd4244 /wd4190 /wd4101 /wd4996 /wd4275 /bigobj -DCAFFE2_BUILD_MAIN_LIB -DONNX_BUILD_MAIN_LIB -O2 /showIncludes /Focaffe2\CMakeFiles\torch.dir\__\torch\csrc\jit\mobile\register_mobile_ops.cpp.obj /Fdcaffe2\CMakeFiles\torch.dir\ /FS -c ..\torch\csrc\jit\mobile\register_mobile_ops.cpp
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\ATen/core/op_registration/op_registration.h(234): error C2971: 'c10::detail::WrapKernelFunction': template parameter 'kernel_func': 'at::relu': a variable with non-static storage duration cannot be used as a non-type argument
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\ATen/core/boxing/kernel_function.h(19): note: see declaration of 'c10::detail::WrapKernelFunction'
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\build\aten\src\ATen/Functions.h(6620): note: see declaration of 'at::relu'
18:58:17 ..\torch\csrc\jit\mobile\register_mobile_ops.cpp(158): note: see reference to function template instantiation 'c10::RegisterOperators::Options &&c10::RegisterOperators::Options::kernel<at::Tensor(const at::Tensor &),at::relu>(c10::TensorTypeId) &&' being compiled
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\ATen/core/op_registration/op_registration.h(234): error C2971: 'c10::detail::WrapKernelFunction': template parameter 'kernel_func': 'at::t': a variable with non-static storage duration cannot be used as a non-type argument
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\ATen/core/boxing/kernel_function.h(19): note: see declaration of 'c10::detail::WrapKernelFunction'
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\build\aten\src\ATen/Functions.h(7423): note: see declaration of 'at::t'
18:58:17 ..\torch\csrc\jit\mobile\register_mobile_ops.cpp(167): note: see reference to function template instantiation 'c10::RegisterOperators::Options &&c10::RegisterOperators::Options::kernel<at::Tensor(const at::Tensor &),at::t>(c10::TensorTypeId) &&' being compiled
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\ATen/core/op_registration/op_registration.h(234): error C2971: 'c10::detail::WrapKernelFunction': template parameter 'kernel_func': 'at::addmm': a variable with non-static storage duration cannot be used as a non-type argument
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\aten\src\ATen/core/boxing/kernel_function.h(19): note: see declaration of 'c10::detail::WrapKernelFunction'
18:58:17 C:\Jenkins\workspace\pytorch-builds\pytorch-win-ws2016-cuda9-cudnn7-py3-build\build\aten\src\ATen/Functions.h(8546): note: see declaration of 'at::addmm'
18:58:17 ..\torch\csrc\jit\mobile\register_mobile_ops.cpp(176): note: see reference to function template instantiation 'c10::RegisterOperators::Options &&c10::RegisterOperators::Options::kernel<at::Tensor(const at::Tensor &,const at::Tensor &,const at::Tensor &,c10::Scalar,c10::Scalar),at::addmm>(c10::TensorTypeId) &&' being compiled
18:58:17 Microsoft (R) C/C++ Optimizing Compiler Version 19.11.25547 for x64
18:58:17 Copyright (C) Microsoft Corporation.  All rights reserved.
```

iseeyuan commented:
@peterjc123 and @ezyang Looking at it. Most likely it's because the decltype registration does not work in windows? @smessmer I'll register them in another way. 

BTW, how to check the last CI results happened before the PR is landed? It showed up in this discussion panel, but I cannot see them when after it's landed.

peterjc123 commented:
@iseeyuan I guess you could use the CI HUD from https://ezyang.github.io/pytorch-ci-hud/build1/pytorch-master.

ezyang commented:
Look at the most recent commit in the history here, there's an X, you can see the result there. It is failing in the PR https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-win-ws2016-cuda9-cudnn7-py3-build/79173/console

iseeyuan commented:
@peterjc123 and @ezyang Thanks for the responses! A fix will come up soon.
